bip-undef-0.mediawiki

```diff
@@ -0,0 +1,113 @@
+* Nodes could be sure they do not connect to a MITM
+* Nodes could allow custom message types (private extensions) to authenticated peers
+
+A simple authentication scheme based on elliptic cryptography will allow peers to identify each other and selective allow access to restricted services or reject the connection if the identity could not be verified.
+
+== Specification ==
+
+The authentication scheme proposed in this BIP uses ECDSA, '''secrets will never be transmitted'''.
+
+'''Authentication initialization must only happen if encrypted channels has been established (according to BIP-???? [1]).'''
+
+The '''encryption-session-ID''' is available once channels are encrypted (according to BIP-???? [1]).
+
+Authentication can be verified by verifying the signature of the encryption-session-ID with a provided "identity-pubkey" (standard ec pubkey).
+
+The authentication state must be kept until the encryption/connection terminates.
+
+=== Known-peers and authorized-peers database ===
+Each peer that supports p2p authentication must provide two user editable databases
+
+# '''known-peers''' contain known identity-pubkeys together with a network identifier (IP & port), similar to the "known-host" file supported by openssh.
+# '''authorized-peers''' contains identity-pubkeys, together with allowed services (example: encryption,filtering,fee-estimation).
+
+=== <code>auth</code> message ===
+A peer can send an authenticate request by sending an <code>auth</code>-message to the remote peer.
+
+{|class="wikitable"
+! Field Size !! Description !! Data type !! Comments
+|-
+| 33bytes || identity-pubkey || comp.-pubkey || The identity pubkey of the requesting peer
+|-
+| 64bytes || signature || normalized comp.-signature || A signature of the encryption-session-ID from the requesting-to-responing communication direction.
+|}
+
+Replay attacks are not possible because the <code>auth</code> message must be sent only if communication encryption according to BIP-???? [1]) has been established. Peers must ignore unencrypted <code>auth</code> messages.
+
+The requesting <code>auth</code> message must be responded with an identical <code>auth</code> message. The response-auth message must contain the identity-pubkey from the responding-peer as well as the encryption-session-ID from the opposit communication direction.
+
+The responding peer can reject the authentication request if the peer was not found in the “authorized-peers” database or if the signature was invalid.
+
+Rejecting the <code>auth</code> request will not reveal the responding peers identity that could lead to fingerprinting the node, however this BIP does not cover protection against fingerprinting the requesting node from the perspective of the responding node.
+
+Responding peers must ignore the requesting peer after a unsuccessfully authentication initialization to avoid resource attacks (banning would lead to fingerprinting of peers that support authentication).
```

Comment:
"Responding peers must ignore the requesting peer after an unsuccessful authentication attempt to avoid resource attacks..."